httpbakery/client.go

```
@@ -456,65 +466,168 @@ func (c *Client) AcquireDischarge(ctx context.Context, cav macaroon.Caveat, payl
 	if cause.Info == nil {
 		return nil, errgo.Notef(err, "interaction-required response with no info")
 	}
+	logger.Infof("got interaction-required error, Info %s", jsonMarshal(cause))
 	// Make sure the location has a trailing slash so that
 	// the relative URL calculations work correctly even when
 	// cav.Location doesn't have a trailing slash.
 	loc := appendURLElem(cav.Location, "")
-	m, err := c.interact(ctx, loc, cause.Info.VisitURL, cause.Info.WaitURL)
+	token, m, err := c.interact(ctx, loc, cause, payload)
+	if err != nil {
+		return nil, errgo.Mask(err, IsDischargeError, IsInteractionError)
+	}
+	if m != nil {
+		// We've acquired the macaroon directly via legacy interaction.
+		return m, nil
+	}
+
+	// Try to acquire the discharge again, but this time with
+	// the token acquired by the interaction method.
+	m, err = c.acquireDischarge(ctx, cav, payload, token)
 	if err != nil {
 		return nil, errgo.Mask(err, IsDischargeError, IsInteractionError)
 	}
 	return m, nil
 }
 
+func jsonMarshal(x interface{}) string {
+	data, _ := json.Marshal(x)
```

interesting, why is it safe to ignore the error here. I'm assuming that if there is some problem it just returns a zero string which is what you'd pust somewhere anyway then please could you add a doc-comment to that effect.

---

This should return the error or panic.

---

I've removed it.